### PR TITLE
CMCL-913: Upgrader prototype

### PR DESCRIPTION
**Pull request by gaborkb**
### Purpose of this PR

https://jira.unity3d.com/browse/CMCL-913
https://jira.unity3d.com/browse/CMCL-914
https://jira.unity3d.com/browse/CMCL-915

Upgrade CM2 to CM3.

### Testing status

- [x] Added an automated test
- [x] Passed all automated tests
- [x] Manually tested 

### Documentation status

- [ ] Updated [CHANGELOG](https://keepachangelog.com/en/1.0.0/)
- [ ] Updated README (if applicable)
- [ ] Commented all public classes, properties, and methods
- [ ] Updated user documentation

### Package version

- [ ] Updated package version

### CODE REVIEW

File: com.unity.cinemachine/Runtime/Behaviours/CinemachineFreeLook.cs [hunk 1]
```diff
@@ -549,7 +549,7 @@ private CinemachineVirtualCamera[] CreateRigs(CinemachineVirtualCamera[] copyFro
         {
             // Invalidate the cache
             mOrbitals = null;
-            float[] softCenterDefaultsV = new float[] { 0.5f, 0.55f, 0.6f };
+            float[] softCenterDefaultsV = new float[] { 0.5f, 0.5f, 0.5f };
```
Comment (glabute) [inline com.unity.cinemachine/Runtime/Behaviours/CinemachineFreeLook.cs:552]: why this change?

File: com.unity.cinemachine/Runtime/Components/CinemachineTrackedDolly.cs [hunk 2]
```diff
@@ -335,5 +335,28 @@ private Vector3 AngularDamping
         private float m_PreviousPathPosition = 0;
         Quaternion m_PreviousOrientation = Quaternion.identity;
         private Vector3 m_PreviousCameraPosition = Vector3.zero;
+
+        internal bool IsUpgradable()
+        {
+            return m_YawDamping == m_RollDamping && m_RollDamping == m_PitchDamping;
```
Comment (glabute) [inline com.unity.cinemachine/Runtime/Components/CinemachineTrackedDolly.cs:341]: Not sure I would bother with this check

File: com.unity.cinemachine/Runtime/Components/CinemachineTrackedDolly.cs [hunk 3]
```diff
@@ -335,5 +335,28 @@ private Vector3 AngularDamping
...
+        }
+
+        internal void Upgrade(CinemachineSplineDolly splineDolly)
+        {
+            splineDolly.m_Damping = new Vector3(m_XDamping, m_YDamping, m_ZDamping);
+            splineDolly.m_AngularDamping = m_RollDamping;
```
Comment (glabute) [inline com.unity.cinemachine/Runtime/Components/CinemachineTrackedDolly.cs:347]: I would use max of yaw/pitch/roll damping

File: com.unity.cinemachine/Runtime/Upgrader/CinemachineUpgraderDriver.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
+#if UNITY_EDITOR
+using UnityEngine;
+
+namespace Cinemachine.Upgrader
+{
+    [ExecuteInEditMode]
+    public class CinemachineUpgraderDriver : MonoBehaviour
```
Comment (glabute) [inline com.unity.cinemachine/Runtime/Upgrader/CinemachineUpgraderDriver.cs:7]: How do you intend this to be used?  Is it just for internal development?

File: com.unity.cinemachine/Runtime/Upgrader/CinemachineUpgrader.cs [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using UnityEditor;
+using UnityEditor.SceneManagement;
+using UnityEngine;
+using UnityEngine.Playables;
+using UnityEngine.SceneManagement;
+using UnityEngine.Timeline;
+
+namespace Cinemachine.Upgrader
```
Comment (glabute) [inline com.unity.cinemachine/Runtime/Upgrader/CinemachineUpgrader.cs:15]: Why is this in Runtime and not in Editor?

File: com.unity.cinemachine/Runtime/Behaviours/CinemachineFreeLook.cs (hunk 1 above)
Comment (gaborkb) [reply com.unity.cinemachine/Runtime/Behaviours/CinemachineFreeLook.cs:552]: Ah, it was for my tests. So that the default freelook I create is upgradable. It should be reverted.

File: com.unity.cinemachine/Runtime/Upgrader/CinemachineUpgraderDriver.cs (hunk 4 above)
Comment (gaborkb) [reply com.unity.cinemachine/Runtime/Upgrader/CinemachineUpgraderDriver.cs:7]: It is only for internal development :)

File: com.unity.cinemachine/Runtime/Upgrader/CinemachineUpgrader.cs (hunk 5 above)
Comment (gaborkb) [reply com.unity.cinemachine/Runtime/Upgrader/CinemachineUpgrader.cs:15]: Should be in Editor.

File: com.unity.cinemachine/Runtime/Behaviours/CinemachineFreeLook.cs (hunk 1 above)
Comment (glabute) [reply com.unity.cinemachine/Runtime/Behaviours/CinemachineFreeLook.cs:552]: It's upgradable as-is, with FreeLookModifier (use Tilt modifier)

Comment (gaborkb) [reply com.unity.cinemachine/Runtime/Behaviours/CinemachineFreeLook.cs:552]: How should I convert? It's a different approach, and the conversion seems complex.
RotateToScreenBounds

File: com.unity.cinemachine/Runtime/Behaviours/CinemachineFreeLookModifier.cs [hunk 6]
```diff
@@ -42,6 +42,15 @@ public interface IModifiableScreenPosition
...
     }
 
+
+    /// <summary>
+    /// Interface for CinemachineComponentBase-derived to allow its bias position to be driven
+    /// </summary>
+    public interface IModifiableBiasPosition
+    {
+        Vector2 Bias { get; set; }
+    }
```
Comment (gaborkb) [inline com.unity.cinemachine/Runtime/Behaviours/CinemachineFreeLookModifier.cs:52]: Maybe add a composer/framingTransposer modifiable that has screen bias soft- and hard zone in one?

File: com.unity.cinemachine/Runtime/Behaviours/CinemachineSmoothPath.cs [hunk 7]
```diff
@@ -95,6 +96,28 @@ public override void InvalidateDistanceCache()
...
+            {
+                spline.Spline.Add(new BezierKnot
+                {
+                    Position = waypoints[i].position,
+                    Rotation = Quaternion.identity,
+                });
+                splineRoll.Roll.Add(new DataPoint<float>(i, waypoints[i].roll));
+            }
+        }
+
```
Comment (gaborkb) [inline com.unity.cinemachine/Runtime/Behaviours/CinemachineSmoothPath.cs:120]: Same as above

File: com.unity.cinemachine/Runtime/Behaviours/CinemachineFreeLookModifier.cs (hunk 6 above)
Comment (glabute) [reply com.unity.cinemachine/Runtime/Behaviours/CinemachineFreeLookModifier.cs:52]: I was thinking that too.  Actually, put the whole screen guides thingy in there, not just ScreenPos and Bias.  Can do it in a separate PR after this one lands.

File: com.unity.cinemachine/Runtime/Behaviours/CinemachineFreeLookModifier.cs [hunk 8]
```diff
@@ -380,6 +389,58 @@ public override void Reset(CinemachineVirtualCameraBase vcam)
             }
         }
     }
+    
+    /// <summary>
+    /// Builtin Freelook modifier for Composer's Screen position. Modifies Screen position at the start of the camera pipeline.
+    /// </summary>
+    public class BiasPositionModifier : ComponentModifier<IModifiableBiasPosition>
```
Comment (glabute) [inline com.unity.cinemachine/Runtime/Behaviours/CinemachineFreeLookModifier.cs:396]: Need to put XML comments everywhere in this class, like in the other ones

File: com.unity.cinemachine/Runtime/Behaviours/CinemachineSmoothPath.cs (hunk 7 above)
Comment (glabute) [reply com.unity.cinemachine/Runtime/Behaviours/CinemachineSmoothPath.cs:120]: yes

File: com.unity.cinemachine/Runtime/Components/CinemachineComposer.cs [hunk 9]
```diff
@@ -341,7 +342,7 @@ internal Rect HardGuideRect
                 m_DeadZoneWidth = Mathf.Min(m_DeadZoneWidth, m_SoftZoneWidth);
                 m_DeadZoneHeight = Mathf.Min(m_DeadZoneHeight, m_SoftZoneHeight);
 
-                Vector2 center = value.center;
+                var center = new Vector2((float)Math.Round(value.center.x, 6), (float)Math.Round(value.center.y, 6));
```
Comment (glabute) [inline com.unity.cinemachine/Runtime/Components/CinemachineComposer.cs:345]: do we still need the Round()?

File: com.unity.cinemachine/Runtime/Core/CinemachinePathBase.cs [hunk 10]
```diff
@@ -375,5 +376,7 @@ private void ResamplePath(int stepsPerSegment)
                 }
             }
         }
+
+        internal abstract void UpgradeTo(out SplineContainer spline);
```
[the remark this answers is not in the document]
Comment (glabute) [reply com.unity.cinemachine/Runtime/Core/CinemachinePathBase.cs:380]: yes

**Review by glabute — APPROVED** [review]

File: com.unity.cinemachine/Runtime/Behaviours/CinemachineFreeLookModifier.cs (hunk 6 above)
Comment (gaborkb) [reply com.unity.cinemachine/Runtime/Behaviours/CinemachineFreeLookModifier.cs:52]: I'll do separately [CMCL-987](https://jira.unity3d.com/browse/CMCL-987)

**Review by glabute — APPROVED** [review]
